Docs: Initial pandoc conversion (#1586)
squid-cache/squid: src/squid.md

> 

Author: Done.

> 

Reviewer: > Francesco: gfm supports - in fact, requires - metadata in the form of [Front Matter](https://jekyllrb.com/docs/front-matter/)

Thank you for that research. Hm... AFAICT, the current contents of this file in this PR does not contain metadata in the form of Front Matter. On the other hand, this file is processed using a "gfm" filter. Is this file missing a "required"  metadata section? Does Pandoc supply its own metadata in such cases (and, if "yes", is that supplied-by-default metadata correct for our use cases?).

Reviewer 2: Let me clarify:
- jeckyll / gfm / github pages requires front matter (this is why all pages in our wiki have it), in some cases there's a plugin specifying defaults which makes it moot
- gfm may allow front matter, but how it is interpreted seems to depend on the implementation

I suppose the only way to know is by trying

Reviewer: > `pandoc` documentation indicates that it supports Front Matter. However when I tried to use it for this manual document it did not work. Only supplying the values via pandoc command line arguments produced the required man syntax correctly.

Was that failed attempt specific to `gfm`? Or did it fail with some other filter formats you have tried during this PR lifetime? I am trying to understand if this lack of Front Matter support (in pandoc context) is specific to the `gfm` filter...

Author: You are right here. It does seem to be fine with specifically the GFM format, but not with others.
Updated in 5f63c13e56

# NAME

**squid** - HTTP web proxy caching server

# SYNOPSIS

> **squid** [-dsCFNRSXY] [-\-foreground] [-l facility] [-f file] [-[au] port] [-k signal] [-n service]

Author: Dropped the '-V' option which is supported by the `getopts(2)` in `src/main.cc` (for Windows native build only) but does not have a switch case to handle it on command line parsing.
TODO remove the useless 'V' option support from Squid code.

---
title: squid
section: 8
---
# NAME

**squid** - HTTP web proxy caching server

# SYNOPSIS

**squid** [-dsCFNRSXY] [-\-foreground] [-l facility] [-f file] [-[au] port] [-k signal] [-n service]

**squid** -[hv]

**squid** -n service -i [-dsX] [-\-foreground] [-l facility]

**squid** -n service -r [-dsX] [-\-foreground] [-l facility]

**squid** -n service -O command [-ds] [-\-foreground] [-l facility]

> 

Reviewer: I assume these three are Windows-specific variants. If they have to be listed here, please move them down, after all cross-platform variants that are going to be much more frequently used.

Author: IMO the fact that they are windows-specific is just a matter of missing implementation for other OS.

The intention here is that the alternatives are listed in alphabetical order of option names.

> **squid** -z [-NS] [-dsX] [-\-foreground] [-l facility] [-f file] [-n service]

Reviewer: This PR does not have to improve the old broken synopsis, but if we are going to do that, let's be consistent about listing options that require an argument. Option `-d` requires an argument but is shown as if it does not.

> 

Reviewer: It looks like this PR is trying to enumerate some or possibly all mutually exclusive ways to run "squid" in synopsis. IMO, there are too many of them and/or the enumeration principles are too complex -- the end result does not work well as a quick summary of how to start Squid. It would also be rather difficult to maintain as developers will forget to update some of these lines when adding options (among other potential problems).

I suggest simplifying the whole thing to just say something like:

```
# SYNOPSIS

**squid** [option]...
```

or

```
# SYNOPSIS

**squid** [options]
```

... and let individual option documentation to detail option compatibility, requirements, and other important caveats.

Author: > It looks like this PR is trying to enumerate some or possibly all mutually exclusive ways to run "squid" in synopsis. 

Yes, this is to match with the typical man.8 style for other tools having mutually exclusive CLI groupings. Sometimes called "modes" of operation for the binary.  see ```man mount``` for one example.

FYI, the problems with your proposal is that it requires a large re-write of the documentation in ```# OPTIONS``` section to clarify which ```[options]``` are relevant to which mode. That kind of change is being left out of scope for this PR.

Author: (sorry I mistakenly edited your post instead of quoting. Have tried to revert to original content.)

Reviewer: > > It looks like this PR is trying to enumerate some or possibly all mutually exclusive ways to run "squid" in synopsis.
 
> Yes, this is to match with the typical man.8 style for other tools having mutually exclusive CLI groupings. ...  see `man mount` for one example.

Glad my impression was correct. Other tools use several "typical" styles; each tool should select what works best for that tool, of course. For example, curl and wget are not using this PR style even though they do have "exclusive CLI groupings". As far as Squid is concerned, I stand by my earlier "the end result does not work well as a quick summary of how to start Squid. It would also be rather difficult to maintain" conclusion. The fact that some tools do use PR approach does not affect my Squid-specific conclusion about it.
 
> FYI, the problems with your proposal is that it requires a large re-write of the documentation in `# OPTIONS` section .... That kind of change is being left out of scope for this PR.

Excluding that kind of change from this PR is perfectly fine _as long as_ this PR does not, at the same time, propose to change the existing manual page synopsis to introduce these problematic "exclusive CLI groupings". A PR that opens a Pandora box has to either close it (i.e. undo the corresponding changes) OR accept change requests (that reshape the corresponding new additions) as in-scope. Pick your poison.

FWIW, I strongly recommend undoing (i.e. going back to the current official synopsis style) as far as this PR is concerned. Let this PR focus on Markdown conversion.

A separate followup PR should deal with these issues (and that PR should also avoid these "exclusive CLI groupings" in the synopsis; I believe we can address "which options are [compatible]" concerns in that PR without too much work).

Author: reverted in 91e2569024

-F
> Don't serve any requests until store is rebuilt.

-h
> Print help message.

-i
> Install as a Windows Service (see **"-n"** option).

-k [ reconfigure | rotate | shutdown | interrupt | kill | debug | check | parse ]
> Parse configuration file, then send signal to running copy
  (except **"-k parse"**) and exit.

-l facility
> Use specified syslog facility. Implies **"-s"**

-n name
> Specify Windows Service name to use for service operations.
  The default is **"Squid"**

-N
> No daemon mode.

-\-foreground
> Parent process does not exit until its children have finished.
  It has no effect with **"-N"** which does not fork/exit at
  startup.

-\-kid roleID
> Play a given SMP kid process role, with a given ID.
  **Do not use this option**. It is meant for the master process
  use only.

-O options
> Set Windows Service Command line options in Registry.

-r
> Remove a Windows Service (see **"-n"** option).

-R
> Do not set **REUSEADDR** on port.

-s
> Enable logging to syslog. Also configurable in
  **@SYSCONFDIR@/squid.conf**

-S
> Double-check swap during rebuild.

-u port
> Specify ICP port number.

-v
> Print version and build details.

-X
> Force full debugging.

-Y

> > Only return **UDP_HIT** or **UDP_MISS_NOFETCH** during fast

Reviewer: How about using backticks instead of **bold** for various constants/names? I think a backtick quote would match the intent better than bold.

```suggestion
> Only return `UDP_HIT` or `UDP_MISS_NOFETCH` during fast
```

In my quick test, both markdown sequences look the same (and have no special formatting) on the screen when reading the generated manual page, but actual `.8` file sequences do differ, so YMMV.

Author: Done in 52c109de98, along with other bold references to option names.

  reload.

-z
> Create missing swap directories and other missing
  [cache_dir](http://www.squid-cache.org/Doc/config/cache_dir/)

>   structures, then exit.

Author: done in 726ce9a7e7

logging.  By using the lightweight Internet Cache Protocols ICP,
HTCP or CARP, Squid caches can be arranged in a hierarchy or
mesh for additional bandwidth savings.

Squid consists of a main server program **squid**, some optional
programs for custom processing and authentication, and some
management and client tools.  When squid starts up, it spawns a
configurable number of helper processes, each of which can
perform parallel lookups.  This reduces the amount of time the
cache waits for results.

Squid is derived from the ARPA funded Harvest Project.

This manual page only lists the command line arguments. For
details on how to configure Squid see the file
**@SYSCONFDIR@/squid.conf.documented**, the Squid
[wiki FAQ](https://wiki.squid-cache.org/) and examples, or the
[configuration manual](http://www.squid-cache.org/Doc/config/).

# OPTIONS

-a port

> Specify HTTP port number where Squid should listen for
  requests, in addition to any
  [http_port](http://www.squid-cache.org/Doc/config/html_port/)
  in **@SYSCONFDIR@/squid.conf**

-C

> Do not catch fatal signals.

-d level

> Write debugging to stderr also.

-f file

> Use the given file instead of @SYSCONFDIR@/squid.conf .

>   If the file name starts with a **!** or **|** then it is

Reviewer: For consistency with other similar quotations:

```suggestion
  If the file name starts with a `!` or `|` then it is
```

  assumed to be an external command or command line. Can for

>   example be used to pre\-process the configuration before it is

Reviewer: ```suggestion
  example be used to preprocess the configuration before it is
```

  being read by Squid. To facilitate this Squid also understands
  the common #line notion to indicate the real source file.

-F

> Don't serve any requests until store is rebuilt.

-h

> Print help message.

-i

> Install as a Windows Service (see `-n` option).

-k [ reconfigure | rotate | shutdown | interrupt | kill | debug | check | parse ]

> Parse configuration file, then send signal to running copy
  (except `-k parse`) and exit.

-l facility

> Use specified syslog facility. Implies `-s`

-n name

> Specify Windows Service name to use for service operations.
  The default is **"Squid"**

-N

> No daemon mode.

> -\-foreground

Reviewer: Just for my own edification, why do we need to escape the second dash here?

Author: IIRC the character `-` (minus/dash) is a special escaping character for the `man(N)` interpreter.

Without the slash the interpreter would convert the double-dash into a single hyphen-Long character. Which is both visually false (looks like a single dash), and cannot be cut-n-pasted by admin who do that.

> Parent process does not exit until its children have finished.
  It has no effect with `-N` which does not fork/exit at
  startup.

-\-kid roleID

> Play a given SMP kid process role, with a given ID.
  **Do not use this option**. It is meant for the master process
  use only.

-O options

> Set Windows Service Command line options in Registry.

-r

> Remove a Windows Service (see `-n` option).

-R

> Do not set **REUSEADDR** on port.

-s

> Enable logging to syslog. Also configurable in

>   **@SYSCONFDIR@/squid.conf**

Reviewer: We did not use bold for this earlier.

```suggestion
  @SYSCONFDIR@/squid.conf
```

or

```suggestion
  `@SYSCONFDIR@/squid.conf`
```

(everywhere).

  work.

squid.conf.default

> Reference copy of the configuration file. Always kept up to
  date with the version of Squid you are using.

> Use this to look up the default configuration settings and
  syntax after upgrading.

squid.conf.documented

> Reference copy of the configuration file. Always kept up to
  date with the version of Squid you are using.

> Use this to read the documentation for configuration options
  available in your build of Squid.

> The [online configuration manual](http://www.squid-cache.org/Doc/config/)
  is also available for a full reference of options.

errorpage.css

> CSS Stylesheet to control display of generated error pages.

> Use this to set any company branding you need, it will apply
  to every language Squid provides error pages for.

Some files also located elsewhere:

mime_table @DEFAULT_MIME_TABLE@

> MIME type mappings for FTP gatewaying

> Can be configured with the
  [mime_table directive](http://www.squid-cache.org/Doc/config/mime_table/)

error_directory @DEFAULT_ERROR_DIR@

> Location of Squid error pages and templates.

# AUTHOR

Squid was written over many years by a changing team of

> developers and maintained in turn by

Reviewer: Squid (as a whole) has been maintained by others as well. The list below and "in turn" phrasing are misleading!
